### PR TITLE
Add community pipeline for semantic guidance for FLUX

### PR DESCRIPTION
**Pull request by Marlon154**
# What does this PR do?

This PR adds the new FluxSemanticGuidancePipeline to enable [semantic guidance (original paper)](https://arxiv.org/abs/2301.12247) capabilities in the Flux model family. Works like already [existing semantic guidance pipelines](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/semantic_stable_diffusion/pipeline_semantic_stable_diffusion.py).
Example on usage is in the doc string, it works with:
- FLUX adapters
- Multiple images per prompt
- Multiple prompts
- CFG with negative prompt
- Without CFG

## Examples

| Prompt | Original | Edit | Edited |
|--------|----------|------|---------|
| A cat holding a sign that says hello world | Original | remove "cat", add "dog" | Edited |
| race car on a track |image-original-2|add "black sports car"|image-edit-2|
| A sports car under a tree.    (with [adapter](https://huggingface.co/aleksa-codes/flux-ghibsky-illustration)) | adapter_orginal|add "apple tree", add "dog sitting on the right side of a car"|adapter_edit|

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).

## Who can review?

@linoytsaban

### CODE REVIEW

**Review by linoytsaban — APPROVED** [review]
great work, thanks @Marlon154!

File: examples/community/pipeline_flux_semantic_guidance.py [hunk 1]
```diff
@@ -0,0 +1,1340 @@
...
+        >>>     edit_momentum_scale=0.3,
+        >>>     edit_mom_beta=0.6,
+        >>>     generator=torch.Generator(device="cuda").manual_seed(6543),
+        >>> ).images[0]
+        >>> image.save("semantic_flux.png")
+        ```
+"""
+
+
+def calculate_shift(
```
Comment (hlky) [inline examples/community/pipeline_flux_semantic_guidance.py:88]:
```suggestion
# Copied from diffusers.pipelines.flux.pipeline_flux.calculate_shift
def calculate_shift(
```

File: examples/community/pipeline_flux_semantic_guidance.py [hunk 2]
```diff
@@ -0,0 +1,1340 @@
...
+        self.vae_scale_factor = 2 ** (len(self.vae.config.block_out_channels) - 1) if getattr(self, "vae", None) else 8
+        # Flux latents are turned into 2x2 patches and packed. This means the latent width and height has to be divisible
+        # by the patch size. So the vae scale factor is multiplied by the patch size to account for this
+        self.image_processor = VaeImageProcessor(vae_scale_factor=self.vae_scale_factor * 2)
+        self.tokenizer_max_length = (
+            self.tokenizer.model_max_length if hasattr(self, "tokenizer") and self.tokenizer is not None else 77
+        )
+        self.default_sample_size = 128
+
+    def _get_t5_prompt_embeds(
```
Comment (hlky) [inline examples/community/pipeline_flux_semantic_guidance.py:232]:
```suggestion
    # Copied from diffusers.pipelines.flux.pipeline_flux.FluxPipeline._get_t5_prompt_embeds
    def _get_t5_prompt_embeds(
```

File: examples/community/pipeline_flux_semantic_guidance.py [hunk 3]
```diff
@@ -0,0 +1,1340 @@
...
+
+        _, seq_len, _ = prompt_embeds.shape
+
+        # duplicate text embeddings and attention mask for each generation per prompt, using mps friendly method
+        prompt_embeds = prompt_embeds.repeat(1, num_images_per_prompt, 1)
+        prompt_embeds = prompt_embeds.view(batch_size * num_images_per_prompt, seq_len, -1)
+
+        return prompt_embeds
+
+    def _get_clip_prompt_embeds(
```
Comment (hlky) [inline examples/community/pipeline_flux_semantic_guidance.py:281]:
```suggestion
    # Copied from diffusers.pipelines.flux.pipeline_flux.FluxPipeline._get_clip_prompt_embeds
    def _get_clip_prompt_embeds(
```

File: examples/community/pipeline_flux_semantic_guidance.py [hunk 4]
```diff
@@ -0,0 +1,1340 @@
...
+        prompt_embeds = prompt_embeds.pooler_output
+        prompt_embeds = prompt_embeds.to(dtype=self.text_encoder.dtype, device=device)
+
+        # duplicate text embeddings for each generation per prompt, using mps friendly method
+        prompt_embeds = prompt_embeds.repeat(1, num_images_per_prompt)
+        prompt_embeds = prompt_embeds.view(batch_size * num_images_per_prompt, -1)
+
+        return prompt_embeds
+
+    def encode_prompt(
```
Comment (hlky) [inline examples/community/pipeline_flux_semantic_guidance.py:325]:
```suggestion
    # Copied from diffusers.pipelines.flux.pipeline_flux.FluxPipeline.encode_prompt
    def encode_prompt(
```

File: examples/community/pipeline_flux_semantic_guidance.py [hunk 5]
```diff
@@ -0,0 +1,1340 @@
...
+            prompt_embeds,
+            pooled_prompt_embeds,
+            editing_prompt_embeds,
+            pooled_editing_prompt_embeds,
+            text_ids,
+            edit_text_ids,
+            enabled_editing_prompts,
+        )
+
+    def encode_image(self, image, device, num_images_per_prompt):
```
Comment (hlky) [inline examples/community/pipeline_flux_semantic_guidance.py:515]:
```suggestion
    # Copied from diffusers.pipelines.flux.pipeline_flux.FluxPipeline.encode_image
    def encode_image(self, image, device, num_images_per_prompt):
```

File: examples/community/pipeline_flux_semantic_guidance.py [hunk 6]
```diff
@@ -0,0 +1,1340 @@
...
+
+        if not isinstance(image, torch.Tensor):
+            image = self.feature_extractor(image, return_tensors="pt").pixel_values
+
+        image = image.to(device=device, dtype=dtype)
+        image_embeds = self.image_encoder(image).image_embeds
+        image_embeds = image_embeds.repeat_interleave(num_images_per_prompt, dim=0)
+        return image_embeds
+
+    def prepare_ip_adapter_image_embeds(
```
Comment (hlky) [inline examples/community/pipeline_flux_semantic_guidance.py:526]:
```suggestion
    # Copied from diffusers.pipelines.flux.pipeline_flux.FluxPipeline.prepare_ip_adapter_image_embeds
    def prepare_ip_adapter_image_embeds(
```

File: examples/community/pipeline_flux_semantic_guidance.py [hunk 7]
```diff
@@ -0,0 +1,1340 @@
...
+
+        ip_adapter_image_embeds = []
+        for i, single_image_embeds in enumerate(image_embeds):
+            single_image_embeds = torch.cat([single_image_embeds] * num_images_per_prompt, dim=0)
+            single_image_embeds = single_image_embeds.to(device=device)
+            ip_adapter_image_embeds.append(single_image_embeds)
+
+        return ip_adapter_image_embeds
+
+    def check_inputs(
```
Comment (hlky) [inline examples/community/pipeline_flux_semantic_guidance.py:557]:
```suggestion
    # Copied from diffusers.pipelines.flux.pipeline_flux.FluxPipeline.check_inputs
    def check_inputs(
```

File: examples/community/pipeline_flux_semantic_guidance.py [hunk 8]
```diff
@@ -0,0 +1,1340 @@
...
+        if negative_prompt_embeds is not None and negative_pooled_prompt_embeds is None:
+            raise ValueError(
+                "If `negative_prompt_embeds` are provided, `negative_pooled_prompt_embeds` also have to be passed. Make sure to generate `negative_pooled_prompt_embeds` from the same text encoder that was used to generate `negative_prompt_embeds`."
+            )
+
+        if max_sequence_length is not None and max_sequence_length > 512:
+            raise ValueError(f"`max_sequence_length` cannot be greater than 512 but is {max_sequence_length}")
+
+    @staticmethod
+    def _prepare_latent_image_ids(batch_size, height, width, device, dtype):
```
Comment (hlky) [inline examples/community/pipeline_flux_semantic_guidance.py:635]:
```suggestion
    # Copied from diffusers.pipelines.flux.pipeline_flux.FluxPipeline._prepare_latent_image_ids
    def _prepare_latent_image_ids(batch_size, height, width, device, dtype):
```

File: examples/community/pipeline_flux_semantic_guidance.py [hunk 9]
```diff
@@ -0,0 +1,1340 @@
...
+        latent_image_id_height, latent_image_id_width, latent_image_id_channels = latent_image_ids.shape
+
+        latent_image_ids = latent_image_ids.reshape(
+            latent_image_id_height * latent_image_id_width, latent_image_id_channels
+        )
+
+        return latent_image_ids.to(device=device, dtype=dtype)
+
+    @staticmethod
+    def _pack_latents(latents, batch_size, num_channels_latents, height, width):
```
Comment (hlky) [inline examples/community/pipeline_flux_semantic_guidance.py:649]:
```suggestion
    # Copied from diffusers.pipelines.flux.pipeline_flux.FluxPipeline._pack_latents
    def _pack_latents(latents, batch_size, num_channels_latents, height, width):
```

File: examples/community/pipeline_flux_semantic_guidance.py [hunk 10]
```diff
@@ -0,0 +1,1340 @@
...
+        latents = latents.view(batch_size, num_channels_latents, height // 2, 2, width // 2, 2)
+        latents = latents.permute(0, 2, 4, 1, 3, 5)
+        latents = latents.reshape(batch_size, (height // 2) * (width // 2), num_channels_latents * 4)
+
+        return latents
+
+    @staticmethod
+    def _unpack_latents(latents, height, width, vae_scale_factor):
```
Comment (hlky) [inline examples/community/pipeline_flux_semantic_guidance.py:657]:
```suggestion
    # Copied from diffusers.pipelines.flux.pipeline_flux.FluxPipeline._unpack_latents
    def _unpack_latents(latents, height, width, vae_scale_factor):
```

File: examples/community/pipeline_flux_semantic_guidance.py [hunk 11]
```diff
@@ -0,0 +1,1340 @@
...
+        self.vae.enable_tiling()
+
+    def disable_vae_tiling(self):
+        r"""
+        Disable tiled VAE decoding. If `enable_vae_tiling` was previously enabled, this method will go back to
+        computing decoding in one step.
+        """
+        self.vae.disable_tiling()
+
+    def prepare_latents(
```
Comment (hlky) [inline examples/community/pipeline_flux_semantic_guidance.py:701]:
```suggestion
    # Copied from diffusers.pipelines.flux.pipeline_flux.FluxPipeline.prepare_latents
    def prepare_latents(
```

File: examples/community/pipeline_flux_semantic_guidance.py [hunk 12]
```diff
@@ -0,0 +1,1340 @@
...
+                        (enabled_editing_prompts, noise_guidance.shape[0]),
+                        device=device,
+                        dtype=noise_guidance.dtype,
+                    )
+                    noise_guidance_edit = torch.zeros(
+                        (enabled_editing_prompts, *noise_guidance.shape),
+                        device=device,
+                        dtype=noise_guidance.dtype,
+                    )
+                    # noise_guidance_edit = torch.zeros_like(noise_guidance)
```
Comment (hlky) [inline examples/community/pipeline_flux_semantic_guidance.py:1173]:
```suggestion
```

File: examples/community/pipeline_flux_semantic_guidance.py [hunk 13]
```diff
@@ -0,0 +1,1340 @@
...
+                        noise_guidance_edit_tmp = torch.where(
+                            torch.abs(noise_guidance_edit_tmp) >= tmp[:, :, None],
+                            noise_guidance_edit_tmp,
+                            torch.zeros_like(noise_guidance_edit_tmp),
+                        )
+
+                        noise_guidance_edit[c, :, :, :] = noise_guidance_edit_tmp
+                        # noise_guidance_edit[c] = noise_guidance_edit_tmp
+
+                        # noise_guidance_edit = noise_guidance_edit + noise_guidance_edit_tmp
```
Comment (hlky) [inline examples/community/pipeline_flux_semantic_guidance.py:1248]:
```suggestion
```

File: examples/community/pipeline_flux_semantic_guidance.py [hunk 14]
```diff
@@ -0,0 +1,1340 @@
...
+                    if len(noise_pred_edit_concepts) > warmup_inds.shape[0] > 0:
+                        concept_weights = concept_weights.to("cpu")  # Offload to cpu
+                        noise_guidance_edit = noise_guidance_edit.to("cpu")
+
+                        concept_weights_tmp = torch.index_select(concept_weights.to(device), 0, warmup_inds)
+                        concept_weights_tmp = torch.where(
+                            concept_weights_tmp < 0, torch.zeros_like(concept_weights_tmp), concept_weights_tmp
+                        )
+                        concept_weights_tmp = concept_weights_tmp / concept_weights_tmp.sum(dim=0)
+                        # concept_weights_tmp = torch.nan_to_num(concept_weights_tmp)
```
Comment (hlky) [inline examples/community/pipeline_flux_semantic_guidance.py:1260]:
```suggestion
```

**Review by hlky — APPROVED** [review]
Thanks!

**Review by hlky — APPROVED** [review]
Thanks for making the changes @Marlon154
